File: LeetCode/python/0053.py
class Solution:
    def maxSubArray(self, nums: list[int]) -> int:
        """
        Time:  O(n)
        Space: O(1)
        """
        prefix, min_prefix, max_sum = 0, 0, float("-inf")

        for num in nums:
            prefix += num

            max_sum = max(max_sum, prefix - min_prefix)
            min_prefix = min(min_prefix, prefix)

        return max_sum

File: LeetCode/golang/0053.go
import "math"

func maxSubArray(nums []int) int {
	/*
	Time:  O(n)
	Space: O(1)
	*/
    curr_sum, best_sum := 0, math.MinInt
    for _, num := range nums {
        curr_sum = max(num, curr_sum + num)
        best_sum = max(best_sum, curr_sum)
    }
    return best_sum
}